### PR TITLE
Add `rabbitmq_user_limits` module to support per-user limits

### PR DESCRIPTION
**Pull request by aitorpazos**
##### SUMMARY

Adds support for per-user limits as they were introduced in RabbitMQ 3.8.10

Fixes: #64 

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
`rabbitmq_user_limits` 

##### ADDITIONAL INFORMATION

I need help handling the different RabbitMQ versions support:

1. I need to update the integration tests RabbitMQ deb to use RabbitMQ >= 3.8.10. Not clear to me what is the process to update those resources.
2. Currently the task exits early if RabbitMQ version is < 3.8.10. Is this ok/enough?

Introduces `rabbitmq_user_limits` task to support per-user limits (`max-connections` and `max-channels`) following a similar implementation as done for `rabbitmq_vhost_limits`.

Usage example:
```yaml
- community.rabbitmq.rabbitmq_user_limits:
    user: guest
    max_connections: 64
    state: present
```

### CODE REVIEW

**Comment by odyssey4me** [comment]
I'd like to implement a range of versions in testing to ensure that new and old versions aren't broken. Let mesee if I can work out the kinks for that and we can rebase this onto it once that's done. I'll endeavour to work on that this within the next two weeks.

**Comment by Andersson007** [comment]
@aitorpazos thanks for the new module!
The main issue now is the broken CI I think, though there are attempts to fix it https://github.com/ansible-collections/community.rabbitmq/pull/75
Hope it'll happen soon.
I'd be happy to review this module once it's done. Thanks for sharing the module with the community!

**Comment by Andersson007** [comment]
I close-reopened to trigger CI

**Comment by Andersson007** [comment]
could you please rebase the PR https://docs.ansible.com/ansible/latest/dev_guide/developing_rebasing.html#rebasing-your-branch

**Comment by Andersson007** [comment]
@aitorpazos please don't change CHANGELOG.rst directly. We have [changelog fragments](https://docs.ansible.com/ansible/latest/community/development_process.html#creating-a-changelog-fragment) for that but this is not the case as new modules / plugins shouldn't have them.

File: CHANGELOG.rst
Comment (Andersson007) [inline CHANGELOG.rst:14]:
```suggestion
```
Please remove this. See my comment about fragments

File: plugins/modules/rabbitmq_user_limits.py
Comment (Andersson007) [inline plugins/modules/rabbitmq_user_limits.py:16]:
```suggestion
  - Manage the state of user limits in RabbitMQ.
```

File: plugins/modules/rabbitmq_user_limits.py
Comment (Andersson007) [inline plugins/modules/rabbitmq_user_limits.py:21]:
```suggestion
      - Name of user to manage limits for.
```
Must be there. Please fix for the other options where it's missed.

File: plugins/modules/rabbitmq_user_limits.py
Comment (Andersson007) [inline plugins/modules/rabbitmq_user_limits.py:54]:
```suggestion
- name: Limit both of the max number of connections and channels on the user 'guest'
  community.rabbitmq.rabbitmq_user_limits:
```
This is the doc requirement, please do same for the other examples

File: plugins/modules/rabbitmq_user_limits.py
Comment (Andersson007) [inline plugins/modules/rabbitmq_user_limits.py:49]:
```suggestion
      choices: [present, absent]
note:
- Supports C(check_mode).
```

File: plugins/modules/rabbitmq_user_limits.py
Comment (Andersson007) [inline plugins/modules/rabbitmq_user_limits.py:98]:
```suggestion
              run_in_check_mode=False,
              split_lines=True):

        if not self._module.check_mode or (self._module.check_mode and run_in_check_mode):
```
is the `run_in_check_mode` arg really necessary? For me personally it looks a bit confusing?

File: plugins/modules/rabbitmq_user_limits.py
Comment (Andersson007) [inline plugins/modules/rabbitmq_user_limits.py:118]:
```suggestion
        # 3.8.x style output
```

File: tests/integration/targets/setup_rabbitmq/tasks/ubuntu.yml
Comment (Andersson007) [inline tests/integration/targets/setup_rabbitmq/tasks/ubuntu.yml:40]:
```suggestion
    deb: https://s3.amazonaws.com/ansible-ci-files/test/integration/targets/setup_rabbitmq/rabbitmq-server_3.8.14-1_all.deb
```
Could you please remove this change also? The guys are going to fix CI with a dedicated PR

File: plugins/modules/rabbitmq_user_limits.py
Comment (aitorpazos) [reply plugins/modules/rabbitmq_user_limits.py]: Fixed

File: plugins/modules/rabbitmq_user_limits.py
Comment (aitorpazos) [reply plugins/modules/rabbitmq_user_limits.py]: Fixed

File: plugins/modules/rabbitmq_user_limits.py
Comment (aitorpazos) [reply plugins/modules/rabbitmq_user_limits.py]: This should be fixed

File: plugins/modules/rabbitmq_user_limits.py
Comment (aitorpazos) [reply plugins/modules/rabbitmq_user_limits.py:49]: Added

File: plugins/modules/rabbitmq_user_limits.py
Comment (aitorpazos) [reply plugins/modules/rabbitmq_user_limits.py]: Fixed

File: tests/integration/targets/setup_rabbitmq/tasks/ubuntu.yml
Comment (aitorpazos) [reply tests/integration/targets/setup_rabbitmq/tasks/ubuntu.yml]: Reverted

File: plugins/modules/rabbitmq_user_limits.py
Comment (aitorpazos) [reply plugins/modules/rabbitmq_user_limits.py:100]: It is a bit confusing to me as well, but I wanted to keep the [pattern used in other modules](https://github.com/ansible-collections/community.rabbitmq/blob/48947b67c000f356f757ad8879aa9f760e46b284/plugins/modules/rabbitmq_vhost.py#L65) so this one wouldn't feel too alien.
Functionally it is needed in order to always run the ` _rabbit_version` regardless of the check mode value.

**Comment by aitorpazos** [comment]
Squashed commits, need to work on those failing tests.

File: plugins/modules/rabbitmq_user_limits.py
Comment (Andersson007) [reply plugins/modules/rabbitmq_user_limits.py:100]: okay but would be nice to refactor it somehow if possible. Can be hard for further folks to figure out what's going on and why.

**Comment by Andersson007** [comment]
> Squashed commits, need to work on those failing tests.

Ok, when it's ready, please let us know

File: plugins/modules/rabbitmq_user_limits.py
Comment (aitorpazos) [reply plugins/modules/rabbitmq_user_limits.py]: I will open a separate PR renaming `run_in_check_mode` to `force_exec_in_check_mode` which should make it much easier to understand.

**Comment by aitorpazos** [comment]
@Andersson007 it should be good to go. Can you review? :pray:

**Review by Andersson007 — COMMENTED** [review]
@aitorpazos thanks for fixing!
Generally LGTM, however I'm not a rabbitmq user and it's hard for me to judge.

Does this work with any of supported rabbitmq versions? If it doesn't, it'd be nice to mention the rabbitmq version since which the feature is supported.

This have CI tests and they passed, so I don't see any things that can say not to merge this.

@odyssey4me , as a main maintainer, what do you think?

@aitorpazos anyway, let's wait a bit and on Wednesday, if there are no objections, we'll merge this. Thank you!

**Comment by aitorpazos** [comment]
On RabbitMQ versions which do not support user limits (ie: `<3.8.10`) the module will fail reporting the reason to the user:
```
if self._version and self._version < Version('3.8.10'):
    self._module.fail_json(changed=False,
                                       msg="User limits are only available for RabbitMQ >= 3.8.10. Detected version: %s" % self._version)
```

**Comment by Andersson007** [comment]
> On RabbitMQ versions which do not support user limits (ie: `<3.8.10`) the module will fail reporting the reason to the user:
> 
> ```
> if self._version and self._version < Version('3.8.10'):
>     self._module.fail_json(changed=False,
>                                        msg="User limits are only available for RabbitMQ >= 3.8.10. Detected version: %s" % self._version)
> ```

Ah, missed that, great, thanks for clarifying!

File: plugins/modules/rabbitmq_user_limits.py
Comment (Andersson007) [inline plugins/modules/rabbitmq_user_limits.py:16]:
```suggestion
  - Manage the state of user limits in RabbitMQ. Supported since RabbitMQ version 3.8.10.
```
How about this or something with the same meaning?

**Review by Andersson007 — APPROVED** [review]
LGTM. Let's wait until Wednesday, then will merge, thanks!

File: tests/integration/targets/setup_rabbitmq/tasks/ubuntu-1.yml
Comment (odyssey4me) [inline tests/integration/targets/setup_rabbitmq/tasks/ubuntu-1.yml:17]: A point of style, but I won't block this PR on it.

This is a list item, so this would execute faster if it was expressed as:

  apt:
    name:
      - apt-transport-https
      - sudo

**Review by odyssey4me — APPROVED** [review]

File: tests/integration/targets/setup_rabbitmq/tasks/ubuntu-1.yml
Comment (Andersson007) [reply tests/integration/targets/setup_rabbitmq/tasks/ubuntu-1.yml:17]: @aitorpazos would you like to change the tests how @odyssey4me 's suggesting?

Comment (aitorpazos) [reply tests/integration/targets/setup_rabbitmq/tasks/ubuntu-1.yml]: Yeah, you are right... pushed

**Comment by Andersson007** [comment]
@aitorpazos thanks for the new module and my congratulations! @odyssey4me thanks for reviewing!

I think we could release the collection soon.

@aitorpazos would be nice to get feedback from a specialist on https://github.com/ansible-collections/community.rabbitmq/pull/83#pullrequestreview-713422377 (and a cople of comments below). Could you please take a look?

**Pull request merged**
